src/source/nats.rs

```diff
@@ -0,0 +1,178 @@
+        SourceManager::start(source, config).await
+    }
+
+    fn default_codec(&self) -> &str {
+        "json"
+    }
+}
+
+pub struct Int {
+    onramp_id: TremorURL,
+    config: Config,
+    subscription: Option<Subscription>,
+    connection: Option<NatsConnection>,
+    origin_uri: EventOriginUri,
+}
+
+impl std::fmt::Debug for Int {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NATS")
+    }
+}
+
+impl Int {
+    fn from_config(uid: u64, onramp_id: TremorURL, config: &Config) -> Self {
+        let config = config.clone();
+        let origin_uri = EventOriginUri {
+            uid,
+            scheme: "tremor-nats".to_string(),
+            host: "not-connected".to_string(),
+            port: None,
+            path: vec![],
+        };
+        Self {
+            onramp_id,
+            config,
+            subscription: None,
+            connection: None,
+            origin_uri,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Source for Int {
+    async fn pull_event(&mut self, _id: u64) -> Result<SourceReply> {
+        if let Some(sub) = &self.subscription {
+            if let Some(msg) = sub.next().await {
+                let mut origin_uri = self.origin_uri.clone();
+                origin_uri.path = vec![msg.subject];
+                let data = msg.data;
+                let mut nats_meta_data = Value::object_with_capacity(1);
+                let mut msg_reply = None;
+                if let Some(reply) = msg.reply {
+                    msg_reply = Some(reply);
+                }
+                let mut msg_headers = None;
+                if let Some(headers) = msg.headers {
+                    let mut key_val: Value = Value::object_with_capacity(headers.len());
+                    for (key, val) in headers.iter() {
+                        let key = String::from(key);
+                        let val: Vec<String> = val.into_iter().map(|x| String::from(x)).collect();
+                        key_val.insert(key, val)?;
+                    }
+                    msg_headers = Some(key_val);
+                }
+                let mut meta_data = Value::object_with_capacity(2);
+                if let Some(msg_reply) = msg_reply {
+                    meta_data.insert("reply", msg_reply)?;
+                }
+                if let Some(msg_headers) = msg_headers {
+                    meta_data.insert("headers", msg_headers)?;
+                }
+                nats_meta_data.insert("nats", meta_data)?;
+                Ok(SourceReply::Data {
+                    origin_uri,
+                    data,
+                    meta: Some(nats_meta_data),
+                    codec_override: None,
+                    stream: 0
+                })
+            } else {
+                Ok(SourceReply::Empty(0))
+            }
+        } else {
+           Ok(SourceReply::StateChange(SourceState::Disconnected))
```

Comment:
when are we getting here and how can we recover from this?

Reply:
Okay, so I think I messed up here. `sub.next()` is a blocking call, which returns `None` if the subscription isn't valid somehow. So we should be returning `Ok(SourceReply::StateChange(SourceState::Disconnected))` instead of `Ok(SourceReply::Empty(0))`. Returning `Ok(SourceReply::Empty(0))` doesn't actually make sense because `sub.next()` is a blocking call (the `async_nats` crate doesn't provide a non blocking method).
I don't think we need to do anything in this `else` block, since `self.subscription` is never being set to `None` anywhere. Regarding recovery, I can't seem to think how to do that since we seem to be just returning an `Ok`. Please do let me know if I'm wrong about this.